### PR TITLE
Add a `*.pyi` file for auto-completion & hints.

### PR DESCRIPTION
**Pull request by Minokori**
#### Description

To enable auto-completion and hints in code editors such as VScode, create a `*.pyi` file. This feature is compatible with `python 3.9` and later versions.

#### Where has this been tested?

  * **Platform(s): Windows Desktop** ...
  * **Python version(s): 3.9.10** ...
  * **Unreal Engine version(s):** ...

#### Possible Drawbacks

It might be incompatible with `python 3.8` or earlier versions.

- - -
This change is [Reviewable](https://reviewable.io/reviews/carla-simulator/carla/7009)

### CODE REVIEW

**Comment by Daraan** [comment]
So far it works with Python 3.7 without greater issues, but only looked into some small aspects. (Thanks a lot btw.!)

**Review by Daraan — CHANGES_REQUESTED** [review]

File: PythonAPI/carla/source/carla/libcarla.pyi [hunk 1]
```diff
@@ -0,0 +1,5403 @@
...
+    # endregion
+
+    # region Getters
+    def get_actor(self, actor_id: int) -> Actor:
+        """Looks up for an actor by ID and returns `None` if not found."""
+
+    def get_actors(self, actor_ids: list = None) -> ActorList:
+        """Retrieves a list of `carla.Actor` elements, either using a list of IDs provided or just listing everyone on stage. If an ID does not correspond with any actor, it will be excluded from the list returned, meaning that both the list of IDs and the list of actors may have different lengths."""
+
+    def get_blueprint_library(self) -> list[BlueprintLibrary]:
```
Comment (Daraan) [inline PythonAPI/carla/source/carla/libcarla.pyi:4740]: This returns just a BlueprintLibrary and not a list.
Wording is a bit ambiguous in in the original documentation.

https://carla.readthedocs.io/en/latest/python_api/#carla.World

**Review by Daraan — CHANGES_REQUESTED** [review]

File: PythonAPI/carla/source/carla/libcarla.pyi [hunk 2]
```diff
@@ -0,0 +1,5403 @@
...
+        """Returns an array with Y pixel coordinate of all the events in the stream."""
+
+    def to_image(self) -> Image:
+        """Converts the image following this pattern: blue indicates positive events, red indicates negative events."""
+    # endregion
+
+    # region Dunder Methods
+    def __getitem__(self, pos: int): ...
+
+    def __iter__(self):
```
Comment (Daraan) [inline PythonAPI/carla/source/carla/libcarla.pyi:1139]: Can this one be improved too?

File: PythonAPI/carla/source/carla/libcarla.pyi [hunk 3]
```diff
@@ -0,0 +1,5403 @@
...
+        """
+        ...
+    # endregion
+
+    # region Dunder methods
+    def __getitem__(self, pos: int) -> Actor:
+        """Returns the actor corresponding to pos position in the list."""
+        ...
+
+    def __iter__(self):
```
Comment (Daraan) [inline PythonAPI/carla/source/carla/libcarla.pyi:518]: `-> List[Actor]`

File: PythonAPI/carla/source/carla/libcarla.pyi [hunk 4]
```diff
@@ -0,0 +1,5403 @@
...
+        """If the `id` attribute is modifiable, changes its value to `value`.
+
+        Args:
+            `id (str)`: The identifier for the attribute that is intended to be changed.\n
+            `value (str)`: The new value for said attribute.\n
+        """
+    # endregion
+
+    # region Dunder Methods
+    def __iter__(self):
```
Comment (Daraan) [inline PythonAPI/carla/source/carla/libcarla.pyi:470]: `-> List[ActorAttribute]:`

**Review by Daraan — CHANGES_REQUESTED** [review]

File: PythonAPI/carla/source/carla/libcarla.pyi [hunk 5]
```diff
@@ -0,0 +1,5403 @@
...
+        """Returns whether this actor is active (`True`) or not (`False`)."""
+        ...
+
+    @property
+    def is_dormant() -> bool:
+        """Returns whether this actor is dormant (`True`) or not (`False`) - the opposite of `is_active`."""
+        ...
+
+    @property
+    def parent() -> ["Actor"]:
```
Comment (Daraan) [inline PythonAPI/carla/source/carla/libcarla.pyi:104]: `[Type]` is not proper. Should be changed to `Optional["Actor"] `or `Union[None, "Actor"]`

**Review by Daraan — CHANGES_REQUESTED** [review]
Update included in my latest pull request

**Comment by Daraan** [comment]
Completed the last todos, and added further fixes as well as a changelog to my pull request on @Minokori's branch.

In my opinion afterwards its ready to merge

**Comment by Daraan** [comment]
Todo: FutureActor currently undocumented class; see #7236

File: PythonAPI/carla/source/carla/libcarla.pyi [hunk 6]
```diff
@@ -0,0 +1,5403 @@
...
+    """Class containing the road information and waypoint managing. Data is retrieved from an OpenDRIVE file that describes the road. A query system is defined which works hand in hand with `carla.Waypoint` to translate geometrical information from the .xodr to natural world points. CARLA is currently working with OpenDRIVE 1.4 standard."""
+
+    # region Instance Variables
+    @property
+    def name() -> str:
+        """The name of the map. It corresponds to the .umap from Unreal Engine that is loaded from a CARLA server, which then references to the .xodr road description."""
+    # endregion
+
+    # region Methods
+    def __init__(self, name: str, xodr_content: str) -> list[Transform]:
```
Comment (Daraan) [inline PythonAPI/carla/source/carla/libcarla.pyi:2319]: `__init__` should not return, and this is additionally wrong imo.

Fixes for all `__init__`s included in the latest pushes to my PR

**Review by Daraan — CHANGES_REQUESTED** [review]

**Comment by Blyron** [comment]
Hey @Minokori All PRs must be against dev, could you correct this please?

**Comment by Blyron** [comment]
If this PR does not go against dev we will need to close it and reject it

**Comment by Daraan** [comment]
I've now opened a rebased and merged PR combining @Minokori and my version. Of course we should prefer Minokori's PR but as it has been kind of stale I now decided to open an up-to-date alternative with #7787 as long as this PR here is not maintained further. I hope this is fine with you all.

**Pull request closed without merging**